src/EGL/libeglerror.c

```diff
@@ -0,0 +1,255 @@
+ *
+ * If only executable code is distributed, then the accompanying
+ * documentation must state that "this software is based in part on the
+ * work of the Khronos Group."
+ *
+ * THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+ */
+
+#include "libeglerror.h"
+
+#include <string.h>
+#include <assert.h>
+
+#include <EGL/egl.h>
+#include <EGL/eglext.h>
+
+#include "glvnd_pthread.h"
+#include "libeglmapping.h"
+#include "libeglcurrent.h"
+#include "utils_misc.h"
+
+enum
+{
+    __EGL_DEBUG_BIT_CRITICAL = 0x1,
+    __EGL_DEBUG_BIT_ERROR = 0x2,
+    __EGL_DEBUG_BIT_WARN = 0x4,
+    __EGL_DEBUG_BIT_INFO = 0x8,
+};
+
+static inline unsigned int DebugBitFromType(GLenum type);
+
+static EGLDEBUGPROCKHR debugCallback = NULL;
+static unsigned int debugTypeEnabled = __EGL_DEBUG_BIT_CRITICAL | __EGL_DEBUG_BIT_ERROR;
+static glvnd_rwlock_t debugLock = GLVND_RWLOCK_INITIALIZER;
+
+static inline unsigned int DebugBitFromType(EGLenum type)
+{
+    assert(type >= EGL_DEBUG_MSG_CRITICAL_KHR &&
+            type <= EGL_DEBUG_MSG_INFO_KHR);
+    return (1 << (type - EGL_DEBUG_MSG_CRITICAL_KHR));
+}
+
+EGLint eglDebugMessageControlKHR(EGLDEBUGPROCKHR callback,
+        const EGLAttrib *attrib_list)
+{
+    struct glvnd_list *vendorList;
+    __EGLvendorInfo *vendor;
+
+    __eglEntrypointCommon();
+
+    __glvndPthreadFuncs.rwlock_wrlock(&debugLock);
+
+    if (callback != NULL) {
+        if (attrib_list != NULL) {
+            unsigned int newEnabled = debugTypeEnabled;
+            int i;
+
+            for (i = 0; attrib_list[i] != EGL_NONE; i += 2) {
+                if (attrib_list[i] >= EGL_DEBUG_MSG_CRITICAL_KHR &&
+                        attrib_list[i] <= EGL_DEBUG_MSG_INFO_KHR) {
+                    if (attrib_list[i + 1]) {
+                        newEnabled |= DebugBitFromType(attrib_list[i]);
+                    } else {
+                        newEnabled &= ~DebugBitFromType(attrib_list[i]);
+                    }
+                } else {
+                    // On error, set the last error code, call the current
+                    // debug callback, and return the error code.
+                    __glvndPthreadFuncs.rwlock_unlock(&debugLock);
+                    __eglReportError(EGL_BAD_ATTRIBUTE, "eglDebugMessageControlKHR", NULL,
+                            "Invalid attribute 0x%04lx", (unsigned long) attrib_list[i]);
+                    return EGL_BAD_ATTRIBUTE;
+                }
+            }
+
+            debugCallback = callback;
+            debugTypeEnabled = newEnabled;
+        }
+    } else {
```

Comment:
This is a little weird? You only set callback when an attribute list is explicitly given, even if it's just &{EGL_NONE}. The spec doesn't say the list can be NULL to have the same effect, but it also doesn't say any error is generated if it _is_ NULL. The only case mentioned is:

> If the &lt;attrib_list&gt; contains an unknown attribute or value the function will return a EGL_BAD_ATTRIBUTE error.

I guess if EGL in general doesn't allow null lists then this is fine, but it surprised me when writing a test.